src/midi/midi_fluidsynth.cpp

```diff
@@ -519,122 +543,146 @@ void MidiHandlerFluidsynth::Close()
 	}
+	if (last_rendered_ms >= now_ms) {
+		return 0;
 	}
+	// Return the number of frames needed to get current again
+	assert(ms_per_frame > 0.0);
+	const auto elapsed_ms = now_ms - last_rendered_ms;
+	const auto frames     = iround(ceil(elapsed_ms / ms_per_frame));
+	last_rendered_ms += (frames * ms_per_frame);
+	return check_cast<uint16_t>(frames);
 }
 
-void MidiHandlerFluidsynth::PlaySysex(uint8_t *sysex, size_t len)
+// The request to play the channel message is placed in the MIDI work FIFO
+void MidiHandlerFluidsynth::PlayMsg(const uint8_t* msg)
 {
-	assert(channel);
-	channel->WakeUp();
+	constexpr auto channel_message_len = 4;
+	std::vector<uint8_t> message(msg, msg + channel_message_len);
+	MidiWork work{std::move(message), GetNumPendingFrames(), MessageType::Channel};
+	work_fifo.Enqueue(std::move(work));
+}
 
-	const char *data = reinterpret_cast<const char *>(sysex);
-	const auto n = static_cast<int>(len);
-	fluid_synth_sysex(synth.get(), data, n, nullptr, nullptr, nullptr, false);
+// The request to play the sysex message is placed in the MIDI work FIFO
+void MidiHandlerFluidsynth::PlaySysex(uint8_t* sysex, size_t len)
+{
+	std::vector<uint8_t> message(sysex, sysex + len);
+	MidiWork work{std::move(message), GetNumPendingFrames(), MessageType::Sysex};
+	work_fifo.Enqueue(std::move(work));
 }
 
-void MidiHandlerFluidsynth::MixerCallBack(uint16_t requested_frames)
+// Apply the channel message to the service
+void MidiHandlerFluidsynth::ApplyChannelMessage(const std::vector<uint8_t>& msg)
 {
-	while (requested_frames) {
-		const auto frames_to_be_played = std::min(GetRemainingFrames(),
-		                                          requested_frames);
-		const auto sample_offset_in_buffer = play_buffer.data() +
-		                                     last_played_frame * 2;
-
-		assert(frames_to_be_played <= play_buffer.size());
-		channel->AddSamples_sfloat(frames_to_be_played,
-		                           sample_offset_in_buffer);
-
-		requested_frames -= frames_to_be_played;
-		last_played_frame += frames_to_be_played;
+	// clang-format off
+	// Ref: https://www.midi.org/specifications-old/item/table-1-summary-of-midi-message
+	const auto action = static_cast<uint8_t>(msg[0] >> 4);
+	const auto chan_id = static_cast<uint8_t>(msg[0] & 0b1111);
+	switch (action) {
+	case 0b1000: fluid_synth_noteoff(         synth.get(), chan_id, msg[1]);                 break;
+	case 0b1001: fluid_synth_noteon(          synth.get(), chan_id, msg[1], msg[2]);         break;
+	case 0b1010: fluid_synth_key_pressure(    synth.get(), chan_id, msg[1], msg[2]);         break;
+	case 0b1011: fluid_synth_cc(              synth.get(), chan_id, msg[1], msg[2]);         break;
+	case 0b1100: fluid_synth_program_change(  synth.get(), chan_id, msg[1]);                 break;
+	case 0b1101: fluid_synth_channel_pressure(synth.get(), chan_id, msg[1]);                 break;
+	case 0b1110: fluid_synth_pitch_bend(      synth.get(), chan_id, msg[1] + (msg[2] << 7)); break;
```

Comment:
Interesting that the MIDI standard you linked lists these as binary strings. I'm more accustomed to the hex notation:

<img width="648" alt="image" src="https://user-images.githubusercontent.com/698770/210298753-1e8d7647-7fa0-48b7-b3e7-171969fc74cf.png">

(source: https://ccrma.stanford.edu/~craig/articles/linuxmidi/misc/essenmidi.html)

Maybe I like it better because this is quite similar to working with trackers where some tracker commands are divided into two 4-bit "nibbles" and you can set these separately quite nicely with the 2-digit hex notation 😄 

In any case, it's a small thing.

Reply:
For now, I'd rather not touch this change in this PR - I've very worried I'll introduce a type (or something) and I don't have an obvious way of exercises or testing this.  But no harm in opening this as a PR and switching them to hex and swapping out the reference docs to match! :rocket:

Reply:
Sure thing, I might do that later.